### PR TITLE
feat: add eth state cache

### LINKED ISSUE #1532: Add async caching service to Eth RPC api
### Describe the feature

it will be the case that the same state (block, evm env) is requested concurrently.
to reduce disk lookups we can add a caching layer on top of the provider:

this will look like the `BackendHandler` used in foundry
https://github.com/foundry-rs/foundry/blob/d09be021460fe2b28fc9a6e338f3db42894d7e21/evm/src/executor/fork/backend.rs#L82

requests send messages to an independent Task that bundles lookups for the same data and an LRU cache for fetched data.

eg:
1. send `GetBlock(oneshot::Sender<Block>)` message to task
2. check LRU cache
3. if not present check waiting area
    * if lookup in progress: add to Vec<Sender>
    * else spawn another lookup task that then sends the result to the task
4. task receives lookup result: respond to all senders in waiting area

This has the benefit that we don't need an `RwLock` around a cache and can be async instead (ideal for RPC) 

### Additional context

ref https://github.com/paradigmxyz/reth/pull/1511#discussion_r1115481605

### PR DESCRIPTION
**Pull request by mattsse**
Closes #1532

~~WIP~~

Adds an async caching layer for eth related data.

See mod docs for details

Needs followup to integrate in EthApi

### CODE REVIEW

File: crates/rpc/rpc/src/eth/cache.rs [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+where
+    LimitBlocks: Limiter<H256, Block>,
+    LimitEnvs: Limiter<H256, (CfgEnv, BlockEnv)>,
+{
+    ///
+    client: Client,
+    /// The LRU cache for full blocks grouped by their hash.
+    full_block_cache: LruMap<H256, Block, LimitBlocks>,
+    /// The LRU cache for revm environments
+    evm_env_cache: LruMap<H256, (CfgEnv, BlockEnv), LimitEnvs>,
```
[the remark this answers is not in the document]
Comment (mattsse) [reply crates/rpc/rpc/src/eth/cache.rs:89]: I believe this requires disk lookups for header:

https://github.com/paradigmxyz/reth/blob/7275f8d9227287e9f863cb53deaaf023fce7ea01/crates/storage/provider/src/providers/mod.rs#L243

[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/rpc/rpc/src/eth/cache.rs]: My point is that that info is in `full_block_cache`, no? So you would not need to do look ups

[the remark this answers is not in the document]
Comment (mattsse) [reply crates/rpc/rpc/src/eth/cache.rs]: yeh, we could check if the block is cached, but still need the total difficulty. which is also a problem for the full block. as of rn, this isn't optimized yet right so `header_td` does a lookup.

but it doesn't make much sense to read the full block just to fill the env.

[the remark this answers is not in the document]
Comment (onbjerg) [reply crates/rpc/rpc/src/eth/cache.rs]: I guess this is more of an optimization, as it is possible you only want the env of the block and not everything else (transactions, ommers) for e.g. `eth_call`. Marking this as resolved

**Review by onbjerg — APPROVED** [review]
LGTM

**Pull request merged**